File: analyzer/stubs/c/libc/assert.c
/****************************************************************************/
/*                   Copyright (C) 2018 The MOPSA Project                   */
/*                                                                          */
/*   This program is free software: you can redistribute it and/or modify   */
/*   it under the terms of the CeCILL license V2.1.                         */
/*                                                                          */
/****************************************************************************/


/*
  libc stub
  based on header from glibc-2.27-r6
*/
#include <assert.h>
#include "mopsa_libc_utils.h"


/*$
 * requires: valid_string(__assertion);
 * requires: valid_string(__file);
 * requires: valid_string(__function);
 * ensures : 1 == 0;
 */
void __assert_fail (const char *__assertion, const char *__file,
                    unsigned int __line, const char *__function);

/*$
 * requires: valid_string(__file);
 * requires: valid_string(__function);
 * ensures: 1 == 0;
 */
void __assert_perror_fail (int __errnum, const char *__file,
                           unsigned int __line, const char *__function);

/*$
 * requires: valid_string(__assertion);
 * requires: valid_string(__file);
 * ensures: 1 == 0;
 */
void __assert (const char *__assertion, const char *__file, int __line);

File: analyzer/languages/python/desugar/assert.ml
(****************************************************************************)
(*                                                                          *)
(* This file is part of MOPSA, a Modular Open Platform for Static Analysis. *)
(*                                                                          *)
(* Copyright (C) 2017-2019 The MOPSA Project.                               *)
(*                                                                          *)
(* This program is free software: you can redistribute it and/or modify     *)
(* it under the terms of the GNU Lesser General Public License as published *)
(* by the Free Software Foundation, either version 3 of the License, or     *)
(* (at your option) any later version.                                      *)
(*                                                                          *)
(* This program is distributed in the hope that it will be useful,          *)
(* but WITHOUT ANY WARRANTY; without even the implied warranty of           *)
(* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *)
(* GNU Lesser General Public License for more details.                      *)
(*                                                                          *)
(* You should have received a copy of the GNU Lesser General Public License *)
(* along with this program.  If not, see <http://www.gnu.org/licenses/>.    *)
(*                                                                          *)
(****************************************************************************)

(** Handling of assert statements. *)

open Mopsa
open Sig.Abstraction.Stateless
open Universal.Ast
open Ast

module Domain =
  struct

    include GenStatelessDomainId(struct
        let name = "python.desugar.assert"
      end)

    let checks = []

    let init _ _ flow = None

    let exec stmt man flow =
      let range = srange stmt in
      match skind stmt with
      (* S⟦ assert(e, msg) ⟧ *)
      | S_py_assert (e, msg)->
         (if man.lattice.is_bottom (Flow.get T_cur man.lattice flow) then
           Flow.add_safe_check Alarms.CHK_PY_ASSERTIONERROR e.erange flow |> Post.return
         else
           man.eval (Utils.mk_builtin_call "bool" [e] range) flow >>$ fun e flow ->
         Flow.join
           man.lattice
           (man.exec (mk_assume e range) flow |> post_to_flow man |> Flow.add_safe_check Alarms.CHK_PY_ASSERTIONERROR e.erange)
           (man.exec (mk_assume (mk_py_not e range) range) flow >>%
              man.exec (Utils.mk_builtin_raise "AssertionError" range) |> post_to_flow man)
         |> Post.return
         )
         |> OptionExt.return

      | _ -> None

    let eval _ _ _ = None


    let ask _ _ _ = None

    let print_expr _ _ _ _ = ()

end

let () =
  register_stateless_domain (module Domain)
